events/tests/test_notifications.py

```diff
@@ -166,6 +168,31 @@ def test_event_publish_notification(
     assert len(mail.outbox) == 5  # 3 children of which one has 3 guardians
 
 
+@pytest.mark.django_db
+def test_event_publish_notification_not_sent_when_publication_fails(
+    snapshot,
+    publisher_api_client,
+    notification_template_event_published_fi,
+    unpublished_event,
+    project,
+):
+    GuardianFactory(language="fi")
+    children = ChildWithGuardianFactory.create_batch(3, project=project)
+    children[1].guardians.set(GuardianFactory.create_batch(3, language="fi"))
+    ChildWithGuardianFactory.create_batch(2, project=ProjectFactory(year=2019))
+
+    event_variables = deepcopy(PUBLISH_EVENT_VARIABLES)
+    event_variables["input"]["id"] = to_global_id("EventNode", unpublished_event.id)
+
+    with mock.patch(
+        "events.models.Event.publish",
+        side_effect=Error("Some error occured"),
+    ):
+        publisher_api_client.execute(PUBLISH_EVENT_MUTATION, variables=event_variables)
+        # No mails sent because publication failed
+        assert len(mail.outbox) == 0
```

Comment:
Using mock.patch like this AFAIK results in the original events.models.Event.publish function not being called at all and only the side_effect Error being raised, tried it out locally using pytest and if confirms this behavior. So, this is testing that if events.models.Event.publish would be a function that does nothing else besides raise an Error then no mail is sent? Maybe overriding the events.models.Event.save function and making it fail would be a better test? i.e.
```python
    with mock.patch(
        "events.models.Event.save",
        side_effect=Error("Some error occured"),
    ):
        publisher_api_client.execute(PUBLISH_EVENT_MUTATION, variables=event_variables)
        # No mails sent because publication failed
        assert len(mail.outbox) == 0
```
What do you think?

Reply:
Nice catch! Changed.